Check and fix integer overflows

kitcambridge opened the pull request:
@franziskuskiefer, PTAL at your convenience.

Commit:
Fix integer overflows in chunking bounds checks.

Commit:
Change `base64url_decode` max length calculation to avoid overflow.

Commit:
Check for overflow in payload and plaintext length calculations.

franziskuskiefer commented on src/base64url.c:
```diff
@@ -71,7 +71,19 @@ ece_base64url_decode(const char* base64, size_t base64Len,
+  // Determine the size of the decoded result. Every 4 bytes of input decode
+  // to 3 bytes of output. The final quantum can be 2 or 3 bytes: 2 bytes
+  // decode to 1 byte, and 3 bytes decode to 2 bytes.
+  size_t maxDecodedLen = (base64Len / 4) * 3;
```
I'd recommend multiplying with 0.75. This might actually yield wrong results, e.g. (50/4)*3 = 37.5, but because 50/4 is floored you'll get 36 as result.

kitcambridge replied:
Hmm, I'm not sure I want to make this a float. I *think* the flooring behavior is OK, but maybe not? If the length is 50, the string holds 12 groups of 4 + 2 at the end. Base64 decodes 4 to 3 and 2 to 1, so (12 * 3) + 1 = 37.

franziskuskiefer replied:
Ah, right you add the necessary bits later. Then this should be fine.

franziskuskiefer commented on src/base64url.c:
```diff
@@ -71,7 +71,19 @@ ece_base64url_decode(const char* base64, size_t base64Len,
+  // decode to 1 byte, and 3 bytes decode to 2 bytes.
+  size_t maxDecodedLen = (base64Len / 4) * 3;
+  size_t remaining = base64Len % 4;
+  if (remaining == 2) {
```
A switch statement would probably be nicer.

kitcambridge replied:
Good call! I'll change this and the block below the loop to use matching switches.

franziskuskiefer commented on src/decrypt.c:
```diff
@@ -91,19 +97,27 @@ ece_decrypt_records(const uint8_t* key, const uint8_t* nonce, uint32_t rs,
-    if (recordEnd > ciphertextLen) {
+    size_t recordEnd;
+    if (rs > ciphertextLen - recordStart) {
+      // This check is equivalent to `recordStart + rs > ciphertextLen`; it's
```
Hm, this might not entirely the same. I'm not entirely sure how this is handled, but ciphertextLen - recordStart could be negative, which would overflow if it's handled as an unsigned. So maybe checking that ciphertextLen > recordStart in addition would make sense.

kitcambridge replied:
I don't think `recordStart` can exceed `ciphertextLen`, the exit condition for the loop is `recordStart < ciphertextLen`. (I should also call this variable `ciphertextStart`, to make it clearer it's an offset into the ciphertext...and match the encryption code).

franziskuskiefer commented on src/decrypt.c:
```diff
@@ -91,19 +97,27 @@ ece_decrypt_records(const uint8_t* key, const uint8_t* nonce, uint32_t rs,
-    if (blockEnd > maxBlockEnd) {
+    size_t dataLen = recordLen - ECE_TAG_LENGTH;
+    size_t blockEnd;
+    if (dataLen > maxBlockEnd - blockStart) {
```
same comment here

kitcambridge replied:
Actually, I don't think this is necessary at all, good catch! `recordLen - ECE_TAG_LENGTH` is sufficient, and we know we have enough space because we checked `maxDecryptedLen(rs, ciphertextLen)` before we started decrypting.

franziskuskiefer reviewed (changes requested):
With the x - y you have the problem of underflows. 
size_t z = x - y might give you something pretty large when x < y.

Commit:
Use matching switches in `ece_base64url_decode`.

Commit:
Clean up names and remove redundant check in `ece_decrypt_records`.

* Use `ciphertextStart` and `plaintextStart` to match
  `ece_webpush_encrypt_plaintext`.
* Add comments to clarify what's what.
* Remove `blockEnd`. We know the full block length from
  `recordLen - ECE_TAG_LENGTH`, and we've already checked that the
  output array can hold the full plaintext.

Commit:
Ensure `size_t` lengths fit into `int`s before casting.

Commit:
Fix overallocation for "aesgcm" plaintexts. Closes #24.

* Factor out max plaintext length calculation into
  `ece_plaintext_max_length`, to match `ece_ciphertext_max_length`.
* Fix an underflow when calculating the length of a truncated
  ciphertext. For example, if the ciphertext is 19 bytes, and
  rs = 18, we don't have enough to form a complete second record,
  and 19 - (ECE_TAG_LENGTH * 2) will underflow.

Commit:
Fix pad length check and simplify loop in `ece_aesgcm_unpad`.

Commit:
Set `keyId` to `NULL` for empty key IDs.

kitcambridge commented:
@franziskuskiefer OK, this should take care of the overflow checks, and the other issues you noticed on mozilla-mobile/firefox-ios#2555. Please take a look when you have a chance.

franziskuskiefer commented on src/decrypt.c:
```diff
@@ -1,29 +1,36 @@
-ece_aes128gcm_max_decrypted_length(uint32_t rs, size_t ciphertextLen) {
-  size_t numRecords = (ciphertextLen / rs) + 1;
+ece_plaintext_max_length(uint32_t rs, size_t padSize, size_t ciphertextLen) {
+  size_t overhead = padSize + ECE_TAG_LENGTH;
```
This could overflow.

kitcambridge replied:
This shouldn't overflow `size_t`. `overhead` will be either 17 (aes128gcm) or 18 (aesgcm), and `padSize` is only used internally. Added assertions for this.

franziskuskiefer commented on src/decrypt.c:
```diff
@@ -1,29 +1,36 @@
+  }
+  size_t numRecords = ciphertextLen / rs;
+  if (ciphertextLen % rs) {
+    // A smaller final record is expected for "aes128gcm", and required for
```
I don't think I understand this comment.

kitcambridge replied:
Updated the comment. The final record is smaller than the `rs`, but it should still have an auth tag.

franziskuskiefer commented on src/decrypt.c:
```diff
@@ -196,41 +224,38 @@ ece_webpush_decrypt(const uint8_t* rawRecvPrivKey, size_t rawRecvPrivKeyLen,
-  *blockLen -= offset;
-  memmove(block, pad, *blockLen);
+  size_t plaintextStart = ECE_AESGCM_PAD_SIZE + padLen;
+  *blockLen -= plaintextStart;
```
Are we certain that `*blockLen  >= plaintextStart` here?

kitcambridge replied:
It should be. I moved `plaintextStart` above the loop, just after we check and fail if `padLen > *blockLen - ECE_AESGCM_PAD_SIZE`.

franziskuskiefer commented on src/encrypt.c:
```diff
@@ -220,22 +226,33 @@ ece_webpush_encrypt_plaintext(
-    if (ciphertextEnd > maxCiphertextLen) {
+    // The full length of the encrypted record, including room for the
+    // plaintext, padding, padding delimiter, and auth tag.
+    size_t recordLen = blockLen + overhead;
```
So `recordLen` is what `ciphertextEnd` was before. I'm not sure if I can see that we made sure that this doesn't overflow.

kitcambridge replied:
I renamed `dataPerBlock` to `maxBlockLen`, to make it clearer that `blockLen` should be <= `maxBlockLen`, and added some comments and assertions. Here's my reasoning:

* `maxBlockLen = rs - overhead`. If `rs = UINT32_MAX`, and `sizeof(size_t) == sizeof(uint32_t)`, `maxBlockLen = SIZE_MAX - (17 || 18)`. Also, we check and fail if `rs <= overhead`, so this shouldn't underflow.
* `blockPadLen = min(max(maxBlockLen - 1, 1), padLen)`, so we know `blockPadLen <= padLen`, and `blockPadLen <= maxBlockLen`. `blockPadLen == maxBlockLen` if `rs = overhead + 1`; otherwise, `blockPadLen < maxBlockLen`.
* `blockPlaintextLen = plaintextEnd - plaintextStart`. `blockPlaintextLen = 0` if `rs = overhead + 1`, or if we're writing a trailer.
* `maxBlockPlaintextLen = maxBlockLen - blockPadLen`.
* `blockPlaintextLen <= maxBlockPlaintextLen`, so we know `blockPlaintextLen + blockPadLen` is safe. (It should add up to `maxBlockLen`, except for the last record, or if we have too much padding).
* `blockLen <= maxBlockLen`, so we know `blockLen + overhead` is safe. (It should add up to `rs`, except for the last record).

franziskuskiefer replied:
Ok, makes sense. Thanks for the explanation :)  Sorry for the delayed response.

kitcambridge replied:
Thanks for being thorough!

franziskuskiefer reviewed (commented):
Some nits. But the changes look mostly good to me and should address my earlier comments.

Commit:
Add assertions and comments for length calculations.

Commit:
Ensure that `plaintextStart >= padLen` in `ece_aesgcm_unpad`.

Commit:
Only allocate extra space for short final blocks and trailers.
